### PR TITLE
Add best performance / first place ranks to profile overlay

### PR DESCRIPTION
**Pull request by jorolf**

### CODE REVIEW

**Comment by peppy** [comment]
What are you basing the model's (`Play`) name and structure on?

**Comment by jorolf** [comment]
Not quite sure if I get what you're trying to say.

File: osu.Game/Overlays/Profile/Sections/Ranks/DrawableScore.cs [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+        {
+            protected override IEnumerable<Vector2> ComputeLayoutPositions()
+            {
+                int count = FlowingChildren.Count();
+                for (int i = 0; i < count; i++)
+                    yield return new Vector2(DrawWidth * i * (count == 1 ? 0 : 1f / (count - 1)), 0);
+            }
+        }
+
+        private class ModIcon : Container
```
Comment (peppy) [inline osu.Game/Overlays/Profile/Sections/Ranks/DrawableScore.cs]: why couldn't you use the existing ModIcon implementation?

File: osu.Game/Overlays/Profile/Sections/Ranks/DrawableScore.cs [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+                        Colour = OsuColour.Gray(84),
+                        TextSize = 18,
+                        Position = new Vector2(0f, 2f),
+                        UseFullGlyphHeight = false,
+                    },
+                };
+            }
+        }
+
+        private class LinkContainer : OsuClickableContainer
```
Comment (peppy) [inline osu.Game/Overlays/Profile/Sections/Ranks/DrawableScore.cs]: Why couldn't you use the existing LinkText implementation? Code duplication is bad.

File: osu.Game/Overlays/Profile/Sections/RanksSection.cs [hunk 3]
```diff
@@ -1,12 +1,199 @@
...
+            public override void Clear(bool disposeChildren)
+            {
+                base.Clear(disposeChildren);
+                shownScores = 0;
+                showMoreText.Show();
+            }
+
+            public void ShowMore()
+            {
+                shownScores = Math.Min(Children.Count, shownScores + 5);
```
Comment (peppy) [inline osu.Game/Overlays/Profile/Sections/RanksSection.cs]: This method could be written a lot nicer (you shouldn't need the `shownScores` field at all).

File: osu.Game/Overlays/Profile/Sections/RanksSection.cs [hunk 4]
```diff
@@ -1,12 +1,199 @@
...
+                int i = 0;
+                foreach(DrawableScore score in Children)
+                    score.FadeTo(i++ < shownScores ? 1 : 0);
+                showMoreText.FadeTo(shownScores == Children.Count ? 0 : 1);
+            }
+
+            private class ShowMoreContainer : OsuClickableContainer
```
Comment (peppy) [inline osu.Game/Overlays/Profile/Sections/RanksSection.cs]: This seems to be the same thing as a clickable link container just with a different action. Code duplication is bad.

**Review by peppy — CHANGES_REQUESTED** [review]
see reviews above

File: osu.Game/Overlays/Profile/Sections/Ranks/DrawableScore.cs (hunk 1 above)
Comment (jorolf) [inline osu.Game/Overlays/Profile/Sections/Ranks/DrawableScore.cs]: image

**Comment by peppy** [comment]
I took the liberty of adding API integration.

Todo:

- Remove redundant code – there are two sections being displayed that are almost identical inside `RanksSection`. These should be extracted into their own container to de-dupe the pagination and display logic.
- Make pagination work.
- Load ranking screen when selecting a score.

Let me know whether you want to attempt these remaining tasks or need any advice on a direction to go to make them work.

**Review by peppy — CHANGES_REQUESTED** [review]
Fix specified issues

**Comment by jorolf** [comment]
The only things thats missing now is loading the ranking screen as I'm not sure how a overlay should push a screen

**Comment by peppy** [comment]
There's already a method in `OsuGame` to perform results loading (`LoadScore(Score s)`). You could try making this `internal` or `public` then calling it using DI on `OsuGame`.

If you want to avoid DI, then you can add an `Action<Score> ScoreDisplayRequested` to` ProfileOverlay` and assign the protected method to it in `OsuGame` itself (this way causes less interdependencies so would probably be preferred).

**Review by peppy — APPROVED** [review]

**Pull request merged**